velox/dwio/common/ColumnVisitors.h

```diff
@@ -815,9 +876,13 @@ class DictionaryColumnVisitor
       if (hasHook) {
         translateByDict(input, numInput, values);
         super::values_.hook().addValues(
-            scatter ? scatterRows + super::rowIndex_
-                    : velox::iota(super::numRows_, super::innerNonNullRows()) +
-                    super::rowIndex_,
+            scatter
+                ? scatterRows + super::rowIndex_
+                : shiftRowIndex(
```

Comment:
You need to pass non-null `scatterRows` in this case and do the shift outside this function

Reply:
This line https://github.com/facebookincubator/velox/blob/main/velox/dwio/parquet/reader/RleBpDataDecoder.h#L146 will pass null `scatterRows` , then pass it to `ColumnVisitor#processRun`.

Reply:
You should allocate then shift the indices there then.  And I think we should make the `rowIndex_` to already include `numScanned` if possible.  Even if it is not possible, we should not have different treatment in common code, need to find a way to confine the hack in Parquet.

Reply:
> You should allocate then shift the indices there then.

`newScatterRows` has allocated in `shiftRowIndex`, and `newScatterRows` is passed to `shiftRowIndex` by reference.

> And I think we should make the rowIndex_ to already include numScanned if possible. Even if it is not possible, we should not have different treatment in common code, need to find a way to confine the hack in Parquet.

This is the same as above, the `current == testRow` check will fail.

Reply:
Which line will fail?

Reply:
This [line ](https://github.com/facebookincubator/velox/blob/main/velox/dwio/common/ColumnVisitors.h#L198)will fail